sdk/java/pulumi/src/main/java/com/pulumi/core/Urn.java

```diff
@@ -1,75 +1,244 @@
 package com.pulumi.core;
 
-import com.google.common.base.Strings;
-import com.pulumi.deployment.Deployment;
-import com.pulumi.resources.Resource;
+import com.pulumi.core.internal.Optionals;
+import com.pulumi.core.internal.annotations.InternalUse;
 
-import javax.annotation.Nullable;
 import javax.annotation.ParametersAreNonnullByDefault;
-import java.util.Arrays;
-import java.util.Objects;
+import java.util.Optional;
+
+import static com.pulumi.core.internal.Objects.require;
+import static com.pulumi.core.internal.Strings.isNonEmptyOrNull;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
 
 /**
  * An automatically generated logical URN, used to stably identify resources. These are created
  * automatically by Pulumi to identify resources. They cannot be manually constructed.
+ * <p>
+ * Each resource registered with the Pulumi engine is logically identified
+ * by its uniform resource name (URN).
+ * A resource’s URN is derived from its type, parent type, and user-supplied name.
+ * </p>
+ * <br>
+ * The structure of a URN is defined by the grammar below:
+ * <pre>
+ * urn = "urn:pulumi:" stack "::" project "::" qualified type name "::" name ;
+ *
+ * stack   = string ;
+ * project = string ;
+ * name    = string ;
+ * string  = (* any sequence of unicode code points that does not contain "::" *) ;
+ *
+ * qualified type name = [ parent type "$" ] type ;
+ * parent type         = type ;
+ *
+ * type       = package ":" [ module ":" ] type name ;
+ * package    = identifier ;
+ * module     = identifier ;
+ * type name  = identifier ;
+ * identifier = unicode letter { unicode letter | unicode digit | "_" } ;
+ * </pre>
  */
 @ParametersAreNonnullByDefault
 public class Urn {
 
-    private Urn() {
-        throw new UnsupportedOperationException("static class");
+    private static final String Prefix = "urn:pulumi:";
+    private static final String PartsSeparator = "::";
+    private static final String TypeSeparator = ":";
+    private static final String ParentSeparator = "$";
+    private static final String ParentSeparatorRegex = "\\$";
+
+    public final String stack;
+    public final String project;
+    public final QualifiedTypeName qualifiedType;
+    public final String name;
+
+    private Urn(String stack, String project, QualifiedTypeName qualifiedType, String name) {
+        this.stack = requireNonNull(stack);
+        this.project = requireNonNull(project);
+        this.qualifiedType = requireNonNull(qualifiedType);
+        this.name = requireNonNull(name);
+    }
+
+    @InternalUse
+    @SuppressWarnings("ConstantConditions") // IntelliJ can't understand the custom validators
+    public static Urn parse(String urn) {
```

Comment:
The validation is based on the grammar and the examples I've gathered from tests in other implementations.

There is a slight risk of a provider violating one of the rules, but IMHO we should fix the provider and or extend the validation rather than have undertested code.